src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrManager.cs

```diff
@@ -0,0 +1,499 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using LibGit2Sharp;
+using Microsoft.DotNet.Darc.Models.VirtualMonoRepo;
+using Microsoft.DotNet.DarcLib.Helpers;
+using Microsoft.Extensions.Logging;
+
+#nullable enable
+namespace Microsoft.DotNet.DarcLib.VirtualMonoRepo;
+
+public class VmrManager : IVmrManager
+{
+    private const string HEAD = "HEAD";
+
+    // Message shown when synchronizing a single commit
+    private const string SingleCommitMessage =
+        """
+        [{name}] Sync {newShaShort}: {commitMessage}
+
+        Original commit: {remote}/commit/{newSha}
+        """;
+
+    // Message shown when synchronizing multiple commits as one
+    private const string SquashCommitMessage =
+        """
+        [{name}] Sync {oldShaShort} → {newShaShort}
+        Diff: {remote}/compare/{oldSha}..{newSha}
+        
+        From: {remote}/commit/{oldSha}
+        To: {remote}/commit/{newSha}
+        
+        Commits:
+        {commitMessage}
+        """;
+
+    private readonly ILogger<VmrManager> _logger;
+    private readonly IProcessManager _processManager;
+    private readonly IRemoteFactory _remoteFactory;
+    private readonly string _vmrPath;
+    private readonly string _reposPath;
+    private readonly string _tagsPath;
+    private readonly string _tmpPath;
+
+    public IReadOnlyCollection<SourceMapping> Mappings { get; }
+
+    public VmrManager(
+        IProcessManager processManager,
+        IRemoteFactory remoteFactory,
+        ILogger<VmrManager> logger,
+        IReadOnlyCollection<SourceMapping> mappings,
+        string vmrPath,
+        string tmpPath)
+    {
+        _logger = logger;
+        _processManager = processManager;
+        _remoteFactory = remoteFactory;
+        _tmpPath = tmpPath;
+        _vmrPath = vmrPath;
+        _reposPath = Path.Join(vmrPath, "src");
+        _tagsPath = Path.Join(_reposPath, ".tags");
+
+        Mappings = mappings;
+    }
+
+    public async Task InitializeVmr(
+        SourceMapping mapping,
+        string? targetRevision,
+        bool ignoreWorkingTree,
+        CancellationToken cancellationToken)
+    {
+        if (File.Exists(Path.Combine(_reposPath, $".{mapping.Name}")))
+        {
+            throw new EmptySyncException($"Repository {mapping.Name} already exists");
+        }
+
+        _logger.LogInformation("Initializing {name}", mapping.Name);
+
+        string clonePath = await PrepareClone(mapping);
+        string patchPath = GetPatchFilePath(mapping);
+
+        cancellationToken.ThrowIfCancellationRequested();
+
+        using var clone = new Repository(clonePath);
+        var commit = GetCommit(clone, targetRevision is null || targetRevision == HEAD ? null : targetRevision);
```

Comment:
could we add () here so it's `(targetRevision is null || targetRevision == HEAD)`, so it's more readable

Reply:
We can but I believe we will get a compilation message saying there's unnecessary parenthesis